File: src/Aiplugs.Elements/Components/Nav/Nav.cs
using Aiplugs.Elements.Extensions;
using Microsoft.AspNetCore.Razor.TagHelpers;

namespace Aiplugs.Elements
{
    [HtmlTargetElement("aiplugs-nav")]
    public class AiplugsNav : TagHelper
    {
        public int FoldThreshold { get; set; } = 768;
        public override void Process(TagHelperContext context, TagHelperOutput output)
        {
            output.TagName = "nav";
            output.TagMode = TagMode.StartTagAndEndTag;
            output.Attributes.Merge("class", "aiplugs-nav");
            output.Attributes.Add("data-controller", "aiplugs-nav");
            output.Attributes.Add("data-aiplugs-nav-threshold", FoldThreshold);
        }
    }

    [HtmlTargetElement(Attributes="[is=aiplugs-nav-fold]")]
    public class AiplugsNavFold : TagHelper
    {
        public override void Process(TagHelperContext context, TagHelperOutput output)
        {
            output.Attributes.Add("data-action", "aiplugs-nav#toggle");
        }
    }
}

File: src/Aiplugs.Elements/Components/Nav/nav.js
AiplugsElements.register('aiplugs-nav', class extends Stimulus.Controller {
    initialize() {
        this.force = this.fold;
        window.addEventListener('resize', () => {
            this.resize();
        })
        this.update();
        this.resize();
    }
    toggle() {
        this.fold = !this.fold;
        this.force = !this.force;
        this.update();
    }
    resize() {
        if (!this.force) {   
            this.fold = window.innerWidth < this.threshold;
            this.update();
        }
    }
    update() {
        this.element.classList.toggle('--fold', this.fold);
        this.element.classList.toggle('--force', this.force);
    }
    get fold() {
        return sessionStorage.getItem('aiplugs-nav-fold') === 'true';
    }
    set fold(value) {
        sessionStorage.setItem('aiplugs-nav-fold', value);
    }
    get force() {
        return sessionStorage.getItem('aiplugs-nav-force') === 'true';
    }
    set force(value) {
        sessionStorage.setItem('aiplugs-nav-force', value);
    }
    get threshold() {
        return parseInt(this.data.get('threshold') || 800);
    }
})